### PR TITLE
Implements limited-input device auth flow, to replace deprecated OOB auth flow.

### PR DESCRIPTION
**Pull request by arcra**
* Motivation for features / changes
The OOB auth flow has been deprecated. We concluded that the limited-input device flow is appropriate for our use case where the uploader runs in an environment where a browser is not available.

* Technical description of changes
Implements a new auth flow which calls an RPC to fetch a device_code, verification_url and user_code, and asks user to visit the verification_url in another device and enter the user_code; then starts polling for the access token after the user authorizes the access from another device.

* Screenshots of UI changes
N/A

* Detailed steps to verify changes work correctly (as executed by you)
- Wrote test script that uses this class, and tested the auth flow and was able to print the credentials.
- Wrote tests.

* Alternate designs / implementations considered
Basically, implementing something similar to this flow or the OOB flow ourselves.

### CODE REVIEW

**Review by bmd3k — COMMENTED** [review]
Sorry, I only had limited time today so I haven't done a complete review nor even tried it out to make sure it works :). Most of the comments are relatively minor issues with documentation, style, and code structure.

File: tensorboard/uploader/auth.py [hunk 1]
```diff
@@ -54,14 +63,30 @@
...
 
+# The config for authenticating using a separate device with a browser, to be
+# used when the environment where this uploader is running does not have a
+# browser available.
+#
+# The client "secret" is considered public, as it's distributed to the client
+# devices where this runs. See:
+# https://developers.google.com/identity/protocols/oauth2/limited-input-device
+#
+# Also see comment in previous constant.
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:79]: Done. Rephrased a bit differently, but it's the same idea. Thanks for the suggestion.

[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py]: Actually, upon second inspection, this wasn't accurate. Ended up consolidating the configs into one variable. Changed it to be directly a dictionary instead of a json that we have to parse before using it. Let me know if there are any concerns with that.

File: tensorboard/uploader/auth.py [hunk 2]
```diff
@@ -149,36 +174,188 @@ def clear(self):
...
-    Returns:
-      CustomInstalledAppFlow: the constructed flow.
-    """
-    return CustomInstalledAppFlow.from_client_config(
-        client_config, scopes=OPENID_CONNECT_SCOPES
-    )
+class _TbUploaderAuthFlow(auth_flows.Flow):
+    """Customized version of an OAuth2 flow for the TB.dev uploader.
 
+    When a browser is not available or the `--auth_force_console` flag is
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py]: Moved this documentation to the authenticate_user method, and left the more complete documentation.

File: tensorboard/uploader/auth.py [hunk 3]
```diff
@@ -149,36 +174,188 @@ def clear(self):
...
+        return self.run_limited_input_device_flow()
+
+    def run_installed_app_flow(self) -> google.oauth2.credentials.Credentials:
+        flow = auth_flows.InstalledAppFlow.from_client_config(
+            self._installed_app_client_config, scopes=self.scopes
+        )
+        return flow.run_local_server(port=0)
+
+    def run_limited_input_device_flow(
+        self, **kwargs
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py]: Removed.

File: tensorboard/uploader/auth.py [hunk 4]
```diff
@@ -149,36 +174,188 @@ def clear(self):
...
+    ) -> google.oauth2.credentials.Credentials:
+        print("Using authentication without a browser.\n")
+        device_response = self._send_device_auth_request()
+        prompt_message = (
+            "Please visit this URL in another device, and enter the provided "
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py]: Done.

File: tensorboard/uploader/auth.py [hunk 5]
```diff
@@ -149,36 +174,188 @@ def clear(self):
...
+        print(prompt_message)
+
+        auth_response = self._poll_for_auth_token(
+            device_code=device_response["device_code"],
+            grant_type=self._limited_input_device_client_config["grant_type"],
+            polling_interval=device_response["interval"],
+            expiration_seconds=device_response["expires_in"],
+        )
+
+        if not auth_response:
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py]: Done.

**Review by bmd3k — COMMENTED** [review]
I've gotten though everything except for the tests. Thanks for your patience!

File: tensorboard/uploader/auth.py [hunk 6]
```diff
@@ -43,24 +46,28 @@
...
+    "installed": {
+        "client_id": "373649185512-8v619h5kft38l4456nm2dj4ubeqsrvh6.apps.googleusercontent.com",
+        "project_id": "hosted-tensorboard-prod",
+        "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+        "token_uri": "https://oauth2.googleapis.com/token",
+        "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+        "client_secret": "pOyAuU2yq2arsM98Bw5hwYtr",
+        "redirect_uris": ["http://localhost"],
+    },
+    "limited-input device": {
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:63]: Done.

File: tensorboard/uploader/auth.py [hunk 7]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
+    """Makes the user authenticate to retrieve auth credentials.
+
+    The default behavior is to use the [installed app flow](
+    http://developers.google.com/identity/protocols/oauth2/native-app), in which
+    a browser is started for the user to authenticate, along with a local web
+    server. The authentication in the browser would produce a redirect response
+    to `localhost` with an authorization code that would then be received by the
+    local web server started here.
+
+    Notably, when the uploaoder is run from a colab notebook, this flow cannot
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:171]: Done.

File: tensorboard/uploader/auth.py [hunk 8]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
+    be used, as colab notebooks are run in an environment where a browser is not
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:172]: I'm not sure, but I rephrased in a less explicit way to avoid making false statements.

File: tensorboard/uploader/auth.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:171]: Done.

File: tensorboard/uploader/auth.py [hunk 9]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
+    available, even tho the machine where user is interacting with such notebook
+    might have a browser available.
+
+    If any of the following is true, a different auth flow will be used:
+    - the flag `--auth_force_console` is set to true, or
+    - a browser is not available (e.g. when running in a colab notebook), or
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:178]: Removed the example.

File: tensorboard/uploader/auth.py [hunk 10]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
+    - a local web server cannot be started
+
+    In this case, a [limited-input device flow](
+    http://developers.google.com/identity/protocols/oauth2/limited-input-device)
+    will be used, in which the user is presented with a URL and a short code
+    that they'd need to use to authenticate and authorize access in a separate
+    browser or device, after which the uploader will poll for access until the
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:185]: Done.

File: tensorboard/uploader/auth.py [hunk 11]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
-    )
 
+    def __init__(self, client_config, scopes):
+        self._client_config = client_config["limited-input device"]
+        self._scopes = scopes
+
+    def run(self) -> google.oauth2.credentials.Credentials:
+        device_response = self._send_device_auth_request()
+        prompt_message = (
+            "Please visit this URL in a browser window (can be on a different "
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:217]: Lgtm. Done.

File: tensorboard/uploader/auth.py [hunk 12]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
+    def _send_device_auth_request(self):
+        device_uri = self._client_config["device_uri"]
+        params = {
+            "client_id": self._client_config["client_id"],
+            "scope": " ".join(self._scopes),
+        }
+        r = requests.post(device_uri, data=params).json()
+        if "device_code" not in r:
+            raise RuntimeError(
+                "Auth service temporarily unavailable, try again later."
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:246]: Done.

File: tensorboard/uploader/auth.py [hunk 13]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
+        token_uri = self._client_config["token_uri"]
+        params = {
+            "client_id": self._client_config["client_id"],
+            "client_secret": self._client_config["client_secret"],
+            "device_code": device_code,
+            "grant_type": self._client_config["grant_type"],
+        }
+        expiration_time = time.time() + expiration_seconds
+        # Error cases documented in
+        # https://developers.google.com/identity/protocols/oauth2/limited-input-device#step-2:-handle-the-authorization-server-response
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:262]: Updated.

File: tensorboard/uploader/auth.py [hunk 14]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
+            resp = requests.post(token_uri, data=params)
+            r = resp.json()
+            if "access_token" in r:
+                return r
+            if "error" in r and r["error"] == "access_denied":
+                raise PermissionError("Auth was denied.")
+            if resp.status_code in {400, 401}:
+                raise ValueError("There must be an error in the request.")
+
+            if "error" in r and r["error"] == "authorization_pending":
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:273]: Done.

File: tensorboard/uploader/auth.py [hunk 15]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
+        expiration_time = time.time() + expiration_seconds
+        # Error cases documented in
+        # https://developers.google.com/identity/protocols/oauth2/limited-input-device#step-2:-handle-the-authorization-server-response
+        while time.time() < expiration_time:
+            resp = requests.post(token_uri, data=params)
+            r = resp.json()
+            if "access_token" in r:
+                return r
+            if "error" in r and r["error"] == "access_denied":
+                raise PermissionError("Auth was denied.")
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:269]: Done.

File: tensorboard/uploader/auth.py [hunk 16]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
+            if resp.status_code in {400, 401}:
+                raise ValueError("There must be an error in the request.")
+
+            if "error" in r and r["error"] == "authorization_pending":
+                time.sleep(polling_interval)
+            elif "error" in r and r["error"] == "slow_down":
+                time.sleep(int(polling_interval * 1.5))
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:276]: Added a comment. I don't really expect this to be present in reality, but since it's a retryable error, I opted to handle it somehow.

I was imagining some strange edge-case in which the subsequent request is sent "too quickly" after the interval, and for some reason the server says we're polling too fast. Perhaps this is unrealistic. It shouldn't happen.

File: tensorboard/uploader/auth.py [hunk 17]
```diff
@@ -149,36 +156,147 @@ def clear(self):
...
+                    "An unexpected error occurred while authenticating."
                 )
-        return self.run_console()
+        raise TimeoutError("Timed out waiting for authorization.")
+
+    def _build_credentials(
+        self, auth_response
+    ) -> google.oauth2.credentials.Credentials:
+
+        expiration_timestamp = datetime.datetime.utcfromtimestamp(
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py:287]: Yeah. I thought it was a bit weird as well, but this is how they build the credentials here:
github.com/googleapis/google-auth-library-python-oauthlib/blob/main/google_auth_oauthlib/helpers.py#L140

File: tensorboard/uploader/auth.py [hunk 18]
```diff
@@ -149,36 +174,188 @@ def clear(self):
...
-      client_config (Mapping[str, Any]): The client configuration in the Google
-          client secrets format.
 
-    Returns:
-      CustomInstalledAppFlow: the constructed flow.
-    """
-    return CustomInstalledAppFlow.from_client_config(
-        client_config, scopes=OPENID_CONNECT_SCOPES
-    )
+class _TbUploaderAuthFlow(auth_flows.Flow):
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth.py]: Ended up removing the base class. I think it was misleading and erroneous, because the base methods don't really make sense (or work, probably) for this flow.

File: tensorboard/uploader/auth_test.py [hunk 19]
```diff
@@ -149,5 +155,305 @@ def test_read_invalid_json_file(self):
...
+            mock.patch.object(auth_flows.InstalledAppFlow, "from_client_config")
+        )
+
+        self.mocked_device_auth_flow = self.enter_context(
+            mock.patch.object(
+                auth, "_LimitedInputDeviceAuthFlow", autospec=True
+            )
+        )
+
+    def test_authenticate_user__no_force_console_override__has_display__uses_installed_app_flow(
```
Comment (bmd3k) [inline tensorboard/uploader/auth_test.py:199]: Trying to get these names to be more concise, I'd recommend just naming this:

test_uses_installed_app_flow(self):

* The test is already called AuthenticateUserTest, so perhaps "authenticate_user" is redundant.
* I also consider the "installed app flow" the happy path so it seems unnecessary to include all the scenarios where it might not be used in the name. It might just be preferable to include a comment in the test about the conditions in which its used.

File: tensorboard/uploader/auth_test.py [hunk 20]
```diff
@@ -149,5 +155,305 @@ def test_read_invalid_json_file(self):
...
+    ):
+        self.os_env_display_fn.return_value = None
+
+        auth.authenticate_user()
+
+        self.mocked_installed_auth_flow_creator_fn.assert_not_called()
+        self.mocked_device_auth_flow.assert_called_once()
+        self.mocked_device_auth_flow.return_value.run.assert_called_once()
+
+    def test_authenticate_user__no_force_console_override__has_display__webbrowser_error__uses_device_flow(
```
Comment (bmd3k) [inline tensorboard/uploader/auth_test.py:226]: Perhaps just name "test_uses_device_flow_on_webbrowser_error(self):"

File: tensorboard/uploader/auth_test.py [hunk 21]
```diff
@@ -149,5 +155,305 @@ def test_read_invalid_json_file(self):
...
+        auth.authenticate_user()
+
+        # "installed app" flow was instantiated and ran, which raised an
+        # exception, so the other flow also ran.
+        self.mocked_installed_auth_flow_creator_fn.assert_called_once()
+        self.assertTrue(fake_auth_flow.run_local_server_was_called)
+        self.mocked_device_auth_flow.assert_called_once()
+        self.mocked_device_auth_flow.return_value.run.assert_called_once()
+
+    def test_authenticate_user__force_console_override__uses_device_flow(self):
```
Comment (bmd3k) [inline tensorboard/uploader/auth_test.py:242]: Perhaps just name "test_uses_device_flow_when_force_console(self):"

File: tensorboard/uploader/auth_test.py [hunk 22]
```diff
@@ -149,5 +155,305 @@ def test_read_invalid_json_file(self):
...
+        )
+        self.mocked_post.side_effect = [
+            self._DEVICE_RESPONSE,
+            auth_pending_response,
+            self._AUTH_GRANTED_RESPONSE,
+        ]
+
+        self.flow.run()
+
+        device_params = {
```
Comment (bmd3k) [inline tensorboard/uploader/auth_test.py:338]: Maybe rename expected_device_params and expected_polling_params?

**Review by bmd3k — APPROVED** [review]
Thanks!

File: tensorboard/uploader/auth_test.py (hunk 19 above)
Comment (arcra) [reply tensorboard/uploader/auth_test.py:199]: Done. Totally agree with the redundant authenticate_user.

For the rest... I think it's not always clear what the "happy path" is or why, and it could even change over time, so I like detailing the conditions for each test instead of assuming or describing one as the happy path. But I guess I agree it can be considerably less readable when trying to include the entire state in the test name, so I made them shorter.

Related to the verbosity of the state, I was trying to avoid using _when_ and _and_, to make them shorter, but with your suggestions I removed some of the state part, and I think I found a reasonable brief name that allows better readability.

Thanks for the review and all the helpful comments!

File: tensorboard/uploader/auth_test.py [hunk 23]
```diff
@@ -149,5 +155,305 @@ def test_read_invalid_json_file(self):
...
+        )
+        self.mocked_installed_auth_flow_creator_fn.return_value = fake_auth_flow
+
+        auth.authenticate_user()
+
+        self.mocked_installed_auth_flow_creator_fn.assert_called_once()
+        self.assertTrue(fake_auth_flow.run_local_server_was_called)
+        self.mocked_device_auth_flow.assert_not_called()
+
+    def test_authenticate_user__no_force_console_override__no_display__uses_device_flow(
```
[the remark this answers is not in the document]
Comment (arcra) [reply tensorboard/uploader/auth_test.py:215]: Done.

File: tensorboard/uploader/auth_test.py (hunk 20 above)
Comment (arcra) [reply tensorboard/uploader/auth_test.py:226]: Updated to something different, but it's a bit shorter and easier to read.

File: tensorboard/uploader/auth_test.py (hunk 21 above)
Comment (arcra) [reply tensorboard/uploader/auth_test.py:242]: Done.

File: tensorboard/uploader/auth_test.py (hunk 22 above)
Comment (arcra) [reply tensorboard/uploader/auth_test.py:338]: Done.
